test/testutilsversion.cpp

```diff
@@ -0,0 +1,154 @@
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include <QTest>
+
+#include "base/global.h"
+#include "base/utils/version.h"
+
+class TestUtilsVersion final : public QObject
+{
+    Q_OBJECT
+    Q_DISABLE_COPY_MOVE(TestUtilsVersion)
+
+public:
+    TestUtilsVersion() = default;
+
+private slots:
+    void testConstructors() const
+    {
+        using TwoDigits = Utils::Version<unsigned char, 2, 1>;
+        QCOMPARE(TwoDigits(), TwoDigits());
+        QCOMPARE(TwoDigits(), TwoDigits(0));
+        QCOMPARE(TwoDigits(), TwoDigits(0, 0));
+
+        using ThreeDigits = Utils::Version<int, 3>;
+        // should not compile:
+        // ThreeDigits(1);
+        // ThreeDigits(1, 2);
+        // ThreeDigits(1, 2, 3, 4);
+        QCOMPARE(ThreeDigits(u"1.2.3"_qs), ThreeDigits(1, 2, 3));
+        QCOMPARE(ThreeDigits(QByteArrayLiteral("1.2.3")), ThreeDigits(1, 2, 3));
+
+#if (QT_VERSION >= QT_VERSION_CHECK(6, 3, 0))
+        QVERIFY_THROWS_EXCEPTION(RuntimeError, ThreeDigits(u""_qs));
+        QVERIFY_THROWS_EXCEPTION(RuntimeError, ThreeDigits(u"random_string"_qs));
+        QVERIFY_THROWS_EXCEPTION(RuntimeError, ThreeDigits(QByteArrayLiteral("")));
+        QVERIFY_THROWS_EXCEPTION(RuntimeError, ThreeDigits(QByteArrayLiteral("random_string")));
+#endif
+    }
+
+    void testVersionComponents() const
+    {
+        const Utils::Version<int, 1> version1 {1};
+        QCOMPARE(version1[0], 1);
+        QCOMPARE(version1.majorNumber(), 1);
+        // should not compile:
+        // version1.minorNumber();
+        // version1.revisionNumber();
+        // version1.patchNumber();
+#if (QT_VERSION >= QT_VERSION_CHECK(6, 3, 0))
+        QVERIFY_THROWS_EXCEPTION(std::out_of_range, version1[1]);
+        QVERIFY_THROWS_EXCEPTION(std::out_of_range, version1[2]);
+#endif
+
+        const Utils::Version<int, 4> version2 {10, 11, 12, 13};
+        QCOMPARE(version2[0], 10);
+        QCOMPARE(version2[1], 11);
+        QCOMPARE(version2[2], 12);
+        QCOMPARE(version2[3], 13);
+        QCOMPARE(version2.majorNumber(), 10);
+        QCOMPARE(version2.minorNumber(), 11);
+        QCOMPARE(version2.revisionNumber(), 12);
+        QCOMPARE(version2.patchNumber(), 13);
+    }
+
+    void testToString() const
+    {
+        using OneMandatory = Utils::Version<int, 2, 1>;
+        QCOMPARE(OneMandatory(u"10"_qs).toString(), u"10"_qs);
+
+        using FourDigits = Utils::Version<int, 4>;
+        QCOMPARE(FourDigits().toString(), u"0.0.0.0"_qs);
```

Comment:
It really looks confusing/inconsistent to me that, e.g., `FourDigits(1)` won't compile but `FourDigits()` compile.

Reply:
>but FourDigits() compile

It is the default constructor. It seems to be required and cannot be removed.

Reply:
> It is the default constructor. It seems to be required and cannot be removed.

But why is it required? It doesn't look like it's designed correctly... Or is it a workaround to be compatible with something else?

Reply:
>But why is it required? 
>Or is it a workaround to be compatible with something else?

It seems Qt will use it. The error originates from qt code.

Reply:
> > But why is it required?
> > Or is it a workaround to be compatible with something else?
> 
> It seems Qt will use it. The error originates from qt code.

Okay, since this is an existing problem, I won't bother you with it. But since you're here, could you please provide the above error message?

Reply:
```
[  0%] Building CXX object src/base/CMakeFiles/qbt_base.dir/qbt_base_autogen/mocs_compilation.cpp.o
In file included from /home/user/qBittorrent/_build/src/base/qbt_base_autogen/O2QWORDHJD/../../../../../src/base/search/searchpluginmanager.h:37,
                 from /home/user/qBittorrent/_build/src/base/qbt_base_autogen/O2QWORDHJD/moc_searchpluginmanager.cpp:10,
                 from /home/user/qBittorrent/_build/src/base/qbt_base_autogen/mocs_compilation.cpp:38:
/home/user/qBittorrent/src/base/utils/version.h: In instantiation of ‘constexpr Utils::Version<T, N, Mandatory>::Version(Other ...) [with Other = {}; typename std::enable_if<conjunction_v<std::is_constructible<T, Other>...>, int>::type <anonymous> = 0; T = short int; long unsigned int N = 2; long unsigned int Mandatory = 2]’:
/usr/include/qt6/QtCore/qmetatype.h:2303:65:   required from ‘static constexpr void (* QtPrivate::QMetaTypeForType<S>::getDefaultCtr())(const QtPrivate::QMetaTypeInterface*, void*) [with S = Utils::Version<short int, 2>; QtPrivate::QMetaTypeInterface::DefaultCtrFn = void (*)(const QtPrivate::QMetaTypeInterface*, void*)]’
/usr/include/qt6/QtCore/qmetatype.h:2371:60:   required from ‘constexpr const QtPrivate::QMetaTypeInterface QtPrivate::QMetaTypeInterfaceWrapper<Utils::Version<short int, 2> >::metaType’
/usr/include/qt6/QtCore/qmetatype.h:2442:12:   required from ‘constexpr const QtPrivate::QMetaTypeInterface* QtPrivate::qMetaTypeInterfaceForType() [with T = Utils::Version<short int, 2>]’
/usr/include/qt6/QtCore/qmetatype.h:2503:61:   required from ‘static constexpr QMetaType QMetaType::fromType() [with T = Utils::Version<short int, 2>]’
/usr/include/qt6/QtCore/qmetatype.h:1622:1:   required from ‘static int QMetaTypeId<QHash<Key, T> >::qt_metatype_id() [with T = QString; U = Utils::Version<short int, 2>]’
/usr/include/qt6/QtCore/qmetatype.h:1146:89:   required from ‘static constexpr int QMetaTypeId2<T>::qt_metatype_id() [with T = QHash<QString, Utils::Version<short int, 2> >]’
/usr/include/qt6/QtCore/qmetatype.h:2344:58:   required from ‘static constexpr void (* QtPrivate::QMetaTypeForType<S>::getLegacyRegister())() [with S = QHash<QString, Utils::Version<short int, 2> >; QtPrivate::QMetaTypeInterface::LegacyRegisterOp = void (*)()]’
/usr/include/qt6/QtCore/qmetatype.h:2380:70:   required from ‘constexpr const QtPrivate::QMetaTypeInterface QtPrivate::QMetaTypeInterfaceWrapper<QHash<QString, Utils::Version<short int, 2> > >::metaType’
/usr/include/qt6/QtCore/qmetatype.h:2494:16:   required from ‘constexpr const QtPrivate::QMetaTypeInterface* QtPrivate::qTryMetaTypeInterfaceForType() [with Unique = qt_meta_stringdata_SearchPluginManager_t; TypeCompletePair = TypeAndForceComplete<const QHash<QString, Utils::Version<short int, 2> >&, std::integral_constant<bool, false> >]’
/usr/include/qt6/QtCore/qmetatype.h:2538:55:   required from ‘constexpr const QtPrivate::QMetaTypeInterface* const qt_incomplete_metaTypeArray [20]<qt_meta_stringdata_SearchPluginManager_t, QtPrivate::TypeAndForceComplete<SearchPluginManager, std::integral_constant<bool, true> >, QtPrivate::TypeAndForceComplete<void, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<const QString&, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<bool, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<void, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<const QString&, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<void, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<const QString&, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<const QString&, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<void, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<const QString&, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<void, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<const QString&, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<void, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<const QString&, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<const QString&, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<void, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<const QHash<QString, Utils::Version<short int, 2, 2> >&, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<void, std::integral_constant<bool, false> >, QtPrivate::TypeAndForceComplete<const QString&, std::integral_constant<bool, false> > >’
/home/user/qBittorrent/_build/src/base/qbt_base_autogen/O2QWORDHJD/moc_searchpluginmanager.cpp:178:1:   required from here
/home/user/qBittorrent/src/base/utils/version.h:62:45: error: static assertion failed: Not enough parameters provided
   62 |             static_assert((sizeof...(Other) >= Mandatory), "Not enough parameters provided");
      |                           ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/home/user/qBittorrent/src/base/utils/version.h:62:45: note: the comparison reduces to ‘(0 >= 2)’
make[2]: *** [src/base/CMakeFiles/qbt_base.dir/build.make:76: src/base/CMakeFiles/qbt_base.dir/qbt_base_autogen/mocs_compilation.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:323: src/base/CMakeFiles/qbt_base.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
AFAIK QHash requires default constructor to be available.

Reply:
> QHash requires default constructor to be available.

Yes.